### PR TITLE
Add network domain service, SaveProviderSubnets method

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch creates the new network domain service layer, by adding the first method SaveProviderSubnets.

It must be noted that previously, on the legacy mongodb state, this was not done atomically: it was done by updating/adding subnets one after the oter not within any transaction or locking. This may have lead to races. This method now makes use of UpsertSubnets which takes a slice of SubnetInfo and perform all the operations within a single transaction.

Also, the tests have been copied/adapted from the legacy mongodb state.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

The domain is not wired-up yet, so unit tests should suffice for the moment:
```
go test github.com/juju/juju/domain/network/... -gocheck.v         
```

## Links

**Jira card:** JUJU-5105

### CODE REVIEW

File: domain/network/service/space.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		subnetsToUpsert = append(subnetsToUpsert, subnetToUpsert)
+
+		// Iterate over fan configs.
+		for _, fan := range fans {
+			_, subnetNet, err := net.ParseCIDR(subnet.CIDR)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			subnetWithDashes := strings.Replace(strings.Replace(subnetNet.String(), ".", "-", -1), "/", "-", -1)
+			id := fmt.Sprintf("%s-%s-%s", subnet.ProviderId, network.InFan, subnetWithDashes)
```
Comment (SimonRichardson) [inline domain/network/service/space.go:96]: Move this to a different method so we can unit-test it.

File: domain/network/service/space.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			if subnetNet.IP.To4() == nil {
+				s.logger.Debugf("%s address is not an IPv4 address", subnetNet.IP)
+				continue
+			}
+			// Compute the overlay segment.
+			overlaySegment, err := network.CalculateOverlaySegment(subnet.CIDR, fan)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if overlaySegment != nil {
```
Comment (SimonRichardson) [inline domain/network/service/space.go:106]: That's a shame that `CalculateOverlaySegment` can return `nil, nil`... 

```suggestion
			if err != nil {
				return errors.Trace(err)
			} else if overlaySegment == nil {
				continue
```

Comment (nvinuesa) [reply domain/network/service/space.go]: done in 7356f2a070

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**